### PR TITLE
Rename WorkloadDefinition to ComponentDefinion for Application

### PR DESCRIPTION
**Pull request by yangsoon**
refer to #1061

Replace Workloaddefinition with Componentdefinition in Application-Controller.

- [x] replace workloaddef with componentdef
- [x] fix related tests

### CODE REVIEW

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller.go [hunk 1]
```diff
@@ -65,6 +67,11 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, nil
 	}
 
+	// if Workload.Type is not empty, means componentdefinition refer to an already existing workloaddefinition
+	if componentDefinition.Spec.Workload.Type != "" {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller.go:71]: we also need to generate openAPI and docs in this case

File: pkg/oam/util/helper.go [hunk 2]
```diff
@@ -415,6 +420,23 @@ func GetGVKFromDefinition(dm discoverymapper.DiscoveryMapper, definitionRef v1al
...
 }
 
+// ConvertWorkloadGVK2Definition help convert a GVK to DefinitionReference
+func ConvertWorkloadGVK2Definition(def v1alpha2.WorkloadGVK) (v1alpha2.DefinitionReference, error) {
+	var reference v1alpha2.DefinitionReference
+	gv, err := schema.ParseGroupVersion(def.APIVersion)
+	if err != nil {
+		return reference, err
+	}
+	resource := strings.ToLower(def.Kind) + "s"
```
Comment (wonderflow) [inline pkg/oam/util/helper.go:430]: you need to use `discoverymapper` to do this convert.

File: pkg/oam/util/test_utils.go [hunk 3]
```diff
@@ -132,6 +132,19 @@ func (matcher ErrorMatcher) NegatedFailureMessage(actual interface{}) (message s
 	return format.Message(actual, "not to equal", expectedError)
 }
 
+// UnMarshalStringToComponentDefinition parse a string to a componentDefinition object
+func UnMarshalStringToComponentDefinition(s string) (*v1alpha2.ComponentDefinition, error) {
```
Comment (wonderflow) [inline pkg/oam/util/test_utils.go:136]: why not just replace the func `UnMarshalStringToWorkloadDefinition`, does it still be used?

Comment (yangsoon) [reply pkg/oam/util/test_utils.go:136]: UnMarshalStringToWorkloadDefinitio will not be used, i will remove it

File: pkg/oam/util/helper.go (hunk 2 above)
Comment (yangsoon) [reply pkg/oam/util/helper.go:430]: Got it, I have fix it

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller.go (hunk 1 above)
Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller.go:71]: Got it

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller_test.go [hunk 4]
```diff
@@ -327,4 +335,164 @@ spec:
 			Expect(k8sClient.Get(ctx, client.ObjectKey{Name: invalidComponentDefinitionName, Namespace: namespace}, gotComponentDefinition)).Should(BeNil())
 		})
 	})
+
+	Context("When the ComponentDefinition only container Workload.Definition, should create a WorkloadDefinition", func() {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller_test.go:339]:
```suggestion
	Context("When the ComponentDefinition only contain Workload.Definition, should create a WorkloadDefinition", func() {
```

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller_test.go [hunk 5]
```diff
@@ -327,4 +335,164 @@ spec:
...
+				return err == nil
+			}, 10*time.Second, time.Second).Should(BeTrue())
+			Expect(wd.Name).Should(Equal(def.Name))
+			Expect(wd.Namespace).Should(Equal(def.Namespace))
+			Expect(wd.Annotations).Should(Equal(def.Annotations))
+			Expect(wd.Spec.Schematic).Should(Equal(def.Spec.Schematic))
+		})
+	})
+
+	Context("When the ComponentDefinition container Workload.Type, shouldn't create a WorkloadDefinition", func() {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller_test.go:416]:
```suggestion
	Context("When the ComponentDefinition contain Workload.Type, shouldn't create a WorkloadDefinition", func() {
```

File: pkg/oam/util/helper.go [hunk 6]
```diff
@@ -77,6 +77,23 @@ const (
...
+
+// WorkloadType describe the workload type of ComponentDefinition
+type WorkloadType string
+
+const (
+	// ComponentDef describe a workload of Defined by ComponentDefinition
+	ComponentDef WorkloadType = "ComponentDef"
+
+	// HELMDef describe a workload refer to HELM
+	HELMDef WorkloadType = "HELMDef"
```
Comment (wonderflow) [inline pkg/oam/util/helper.go:93]:
```suggestion
	HELMDef WorkloadType = "HelmDef"
```

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller_test.go (hunk 5 above)
Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller_test.go:416]: sorry, I will fix this grammar error

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller_test.go (hunk 4 above)
Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller_test.go:339]: sorry, I will fix this grammar error

File: pkg/oam/util/helper.go (hunk 6 above)
Comment (yangsoon) [reply pkg/oam/util/helper.go:93]: Got it

File: pkg/oam/util/template.go [hunk 7]
```diff
@@ -42,20 +42,23 @@ func GetScopeGVK(ctx context.Context, cli client.Reader, dm discoverymapper.Disc
 func LoadTemplate(ctx context.Context, cli client.Reader, key string, kd types.CapType) (*Template, error) {
 	switch kd {
 	case types.TypeWorkload:
```
Comment (wonderflow) [inline pkg/oam/util/template.go:44]: TypeComponentDefinition?

Comment (yangsoon) [reply pkg/oam/util/template.go:44]: ok, I will fix it

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	var workloadName = h.cd.Name
+	if h.cd.Spec.Workload.Type != "" {
+		workloadType = util.ReferWorkload
+		workloadName = h.cd.Spec.Workload.Type
+	}
+	if h.cd.Spec.Schematic != nil && h.cd.Spec.Schematic.HELM != nil {
+		workloadType = util.HELMDef
+	}
+	wd := new(v1alpha2.WorkloadDefinition)
+	err := h.Get(ctx, client.ObjectKey{Namespace: h.cd.Namespace, Name: workloadName}, wd)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go:50]: if err !=nil{
   notfound 
}
...

Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go:50]: Got it, I will modify this part of the code logic

File: apis/types/capability.go [hunk 9]
```diff
@@ -80,6 +80,8 @@ type Installation struct {
 type CapType string
 
 const (
+	// TypeComponentDefinition represents OAM ComponentDefinition
+	TypeComponentDefinition CapType = "componentDefinition"
```
Comment (wonderflow) [inline apis/types/capability.go:84]:
```suggestion
	TypeComponentDefinition CapType = "component"
```

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+	if h.cd.Spec.Schematic != nil && h.cd.Spec.Schematic.HELM != nil {
+		workloadType = util.HELMDef
+	}
+
+	wd := new(v1alpha2.WorkloadDefinition)
+	err := h.Get(ctx, client.ObjectKey{Namespace: h.cd.Namespace, Name: workloadName}, wd)
+	if err != nil {
+		switch workloadType {
+		case util.ReferWorkload:
+			klog.Infof("ComponentDefinition %s refer to wrong Workload", h.cd.Name)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go:57]:
```suggestion
			klog.Infof("ComponentDefinition get the referred workloaddefinition %s err %w", h.cd.Name,err)
```

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+			return workloadType, err
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go:58]:
```suggestion
			return workloadType, fmt.Errorf("ComponentDefinition get the referred workloaddefinition %s err %w", h.cd.Name,err)
```

**Review by wonderflow — APPROVED** [review]
Good job!

File: apis/types/capability.go (hunk 9 above)
Comment (yangsoon) [reply apis/types/capability.go:84]: Got it.

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go (hunk 10 above)
Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go:57]: Got it

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go (hunk 11 above)
Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/handler.go:58]: Got it

**Pull request merged**